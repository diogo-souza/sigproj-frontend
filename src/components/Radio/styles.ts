import styled, { css } from 'styled-components';
import { RadioProps } from '.';

export const Label = styled.label<Pick<RadioProps, 'labelColor'>>`
  ${({ theme, labelColor }) => css`
    cursor: pointer;
    display: flex;
    align-items: center;
    > span {
      padding-left: ${theme.spacings.xxsmall};
      color: ${theme.colors[labelColor!]};
      line-height: 1.8rem;
    }
  `}
`;

export const Input = styled.input<Pick<RadioProps, 'fillColor'>>`
  ${({ theme, fillColor }) => css`
    display: flex;
    align-items: center;
    justify-content: center;
    appearance: none;
    width: 1.8rem;
    height: 1.8rem;
    border: 0.2rem solid ${theme.colors[fillColor!]};
    border-radius: 50%;
    background: transparent;
    transition: background 0.1s ease-in-out;
    outline: none;
    cursor: pointer;
    &:focus {
      box-shadow: 0 0 0.5rem ${theme.colors[fillColor!]};
    }
    &:before {
      content: '';
      width: 0.8rem;
      height: 0.8rem;
      border-radius: 50%;
      background: ${theme.colors[fillColor!]};
      transition: opacity 0.1s ease-in-out;
      opacity: 0;
      position: absolute;
    }
    &:checked {
      &:before {
        opacity: 1;
      }
    }
  `}
`;
