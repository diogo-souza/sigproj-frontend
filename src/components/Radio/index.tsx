import React, { InputHTMLAttributes, useState } from 'react';
import * as S from './styles';

type RadioValue = string | ReadonlyArray<string> | number;

export type RadioProps = {
  label: string;
  labelFor: string;
  labelColor?: 'title' | 'white';
  fillColor?: 'primary' | 'secondary' | 'tertiary';
  name: string;
  value: RadioValue;
  onCheck?: (value?: RadioValue) => void;
} & InputHTMLAttributes<HTMLInputElement>;

const Radio: React.FC<RadioProps> = ({
  onCheck,
  label,
  labelFor,
  labelColor = 'title',
  value,
  fillColor = 'secondary',
  ...props
}: RadioProps) => {
  const onChange = () => {
    // eslint-disable-next-line no-unused-expressions
    !!onCheck && onCheck(value);
  };

  return (
    <S.Label htmlFor={labelFor} labelColor={labelColor}>
      <S.Input
        id={labelFor}
        onChange={onChange}
        type="radio"
        value={value}
        fillColor={fillColor}
        {...props}
      />
      <span>{label}</span>
    </S.Label>
  );
};
export default Radio;
