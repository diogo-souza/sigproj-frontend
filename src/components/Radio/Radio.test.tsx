import { screen, waitFor } from '@testing-library/react';
import { renderWithTheme } from 'utils/tests/helpers';
import theme from 'styles/themes';

import userEvent from '@testing-library/user-event';
import Radio from '.';

describe('<Radio />', () => {
  it('should render with label (white)', () => {
    renderWithTheme(
      <Radio
        name="anyName"
        value="anyValue"
        label="Radio"
        labelFor="check"
        labelColor="titleWhite"
      />,
    );

    const label = screen.getByText('Radio');
    expect(label).toBeInTheDocument();
    expect(label).toHaveStyle({ color: theme.colors.white });
  });

  it('should render with label (title)', () => {
    renderWithTheme(
      <Radio
        name="anyName"
        value="anyValue"
        label="Radio"
        labelFor="Teste"
        labelColor="titleBlack"
      />,
    );

    const label = screen.getByText('Radio');
    expect(label).toBeInTheDocument();
    expect(label).toHaveStyle({ color: theme.colors.titleBlack });
  });

  it('should dispatch onCheck when label status changes', async () => {
    const onCheck = jest.fn();
    renderWithTheme(
      <Radio
        name="anyName"
        value="anyValue"
        label="Radio"
        labelFor="Radio"
        onCheck={onCheck}
      />,
    );

    expect(onCheck).not.toHaveBeenCalled();

    userEvent.click(screen.getByLabelText('Radio'));
    await waitFor(() => {
      expect(onCheck).toHaveBeenCalledTimes(1);
    });
    expect(onCheck).toHaveBeenCalledWith('anyValue');
  });

  it('Should be accessible with tab', () => {
    renderWithTheme(
      <Radio name="anyName" value="anyValue" label="Radio" labelFor="Radio" />,
    );

    const radio = screen.getByLabelText('Radio');

    expect(document.body).toHaveFocus();

    userEvent.tab();

    expect(radio).toHaveFocus();
  });
});
