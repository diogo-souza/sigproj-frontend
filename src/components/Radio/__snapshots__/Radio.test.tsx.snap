// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Radio /> should render with label (white) 1`] = `
.c0 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > span {
  padding-left: 0.8rem;
  color: #FFFFFF;
  line-height: 1.8rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  width: 1.8rem;
  height: 1.8rem;
  border: 0.2rem solid #159CAE;
  border-radius: 50%;
  background: transparent;
  -webkit-transition: background 0.1s ease-in-out;
  transition: background 0.1s ease-in-out;
  outline: none;
  cursor: pointer;
}

.c1:focus {
  box-shadow: 0 0 0.5rem #159CAE;
}

.c1:before {
  content: '';
  width: 0.8rem;
  height: 0.8rem;
  border-radius: 50%;
  background: #159CAE;
  -webkit-transition: opacity 0.1s ease-in-out;
  transition: opacity 0.1s ease-in-out;
  opacity: 0;
  position: absolute;
}

.c1:checked:before {
  opacity: 1;
}

<label
  class="c0"
  for="check"
>
  <input
    class="c1"
    id="check"
    name="anyName"
    type="radio"
    value="anyValue"
  />
  <span>
    Radio
  </span>
</label>
`;
